samza-api/src/main/java/org/apache/samza/runtime/ApplicationRunner.java

```diff
@@ -120,7 +63,13 @@ public void waitForFinish() {
    * @return true - application finished before timeout
    *         false - otherwise
    */
-  public boolean waitForFinish(Duration timeout) {
-    throw new UnsupportedOperationException(getClass().getName() + " does not support timed waitForFinish.");
-  }
+  boolean waitForFinish(Duration timeout);
+
+  /**
+   * Add a set of customized {@link MetricsReporter}s in the application
+   *
+   * @param metricsReporters the map of customized {@link MetricsReporter}s objects to be used
+   */
+  void addMetricsReporters(Map<String, MetricsReporter> metricsReporters);
```

Comment:
I am trying to keep all objects registered via ApplicationDescriptors to be serializable s.t. it can be invoked once in describe and serialized. In addition, MetricsReporters are only created in local processes (i.e. standalone or container). We don't use it at all in the ApplicationRunner.main(). It might be better to wire it with ProcessorLifecycleListenerFactory and accessed via the general context as in your design. Thoughts?

Reply:
Sure. The main concern I had was backward compatibility. However, we are class loading this factory anyways. I will move it to ApplicationDescriptor.

Reply:
There is an issue in instantiating StreamProcessor with the customized MetricsReporter instances created from MetricsReporterFactory. Currently, MetricsReporterFactory.getMetricsReporter() requires the processorId is know before we start the StreamProcessor, which is a problem (SAMZA-1835). I will create a follow up PR to address this issue.